examples/perplexity/perplexity.cpp

```diff
@@ -462,7 +462,13 @@ void hellaswag_score(llama_context * ctx, const gpt_params & params) {
         for (size_t ending_idx = 1; ending_idx < 4; ending_idx++) {
 
             // Tokenize the query
-            query_embd = ::llama_tokenize(ctx, hs_data[task_idx].ending[ending_idx], false);
+            // SPM tokenizer: Do not tokenize the starting space in the ending since it is always added by the tokenizer
+            if (is_spm) {
+                query_embd = ::llama_tokenize(ctx, hs_data[task_idx].ending[ending_idx].substr(1,hs_data[task_idx].ending[ending_idx].size()-1), false);
+            } else {
+                query_embd = ::llama_tokenize(ctx, hs_data[task_idx].ending[ending_idx], false);
+            }
```

Comment:
I don't think this is correct. We rather have to remove the leading space from here:

https://github.com/ggerganov/llama.cpp/blob/730d9c681e339b76407659344e5a2cd50af7d7d5/examples/perplexity/perplexity.cpp#L395

I've missed that when looking for the places where we manually add leading spaces.

Reply:
it is correct and works. the context and endings are concatenated before tokenization further down

Reply:
the BPE tokenizer do not add spaces and we need one space

Reply:
Yes - it gives the correct results, but it is the wrong way to do it. You are special casing the SPM tokenizer this way.
Instead, the general solution is to never prefix the text that you want to tokenize. I don't think any other codebase does something like

```python
text = ' ' + text
tokenizer.encode(text)
```

Regardless of the tokenizer used, the correct code is

```
tokenizer.encode(text)
```

Reply:
When doing hellaswag we need to do this since we need to know exactly which token the endings start. This is where the logprobs we need starts

Reply:
We are evaluating the the whole sentences (context+space+ending), but calculate the scores only from the logprobs of the endings.

Reply:
A better way would be to have a param to llama_tokenize `prepend_space` like the one to prepend BOS

Reply:
But we could do it the other way around. Remove the space from the endings and add a space to the BPE instead of subtracting it from the SPM.

Reply:
I didn't notice that this branch was before @ikawrakow's fix, so just updated to latest `master`

The #2805 change is the correct way to compute the Hellaswag, because this way we don't implement special logic for different tokenizers